### PR TITLE
WIP: feat(chat-ui): Support open chat in editor / open chat in new window

### LINKED ISSUE #2944: Support open chat in editor / open chat in new window for Tabby Chat
**Please describe the feature you want**

image

GitHub Copilot supports opening a `Chat` as an individual window or as a separate tab, which is also a feature that can be implemented for Tabby Chat functionality.

**Additional context**
1. https://demo.tabbyml.com/files/github/TabbyML/tabby/-/blob/77515f9c83f32e0db84d454cad671e4a90a032e7/clients/vscode/package.json#L402
2. https://code.visualstudio.com/api/extension-guides/webview (createWebviewPanel)

---
Please reply with a 👍 if you want this feature.

### PR DESCRIPTION
**Pull request by antimonyGu**
close #2944

### CODE REVIEW

**Comment by antimonyGu** [comment]
Hi, all. Do we have any test case to make sure the refactor of **ChatViewProvider** is correct?

File: clients/vscode/src/chat/chatPanel.ts [hunk 1]
```diff
@@ -25,7 +25,7 @@ export function createThreadFromWebview<Self = Record<string, never>, Target = R
   );
 }
 
-export function createClient(webview: WebviewView, api: ClientApi): ServerApi {
+export function createClient(webview: WebviewView | WebviewPanel, api: ClientApi): ServerApi {
```
Comment (wsxiaoys) [inline clients/vscode/src/chat/chatPanel.ts:28]:
```suggestion
export function createClient(webview: Webview, api: ClientApi): ServerApi {
```

File: clients/vscode/src/chat/ChatViewProvider.ts
Comment (wsxiaoys) [inline clients/vscode/src/chat/ChatViewProvider.ts:1]: rename filename (and classname) to `ChatSideViewProvider`

File: clients/vscode/src/chat/BaseChatView.ts [hunk 2]
```diff
@@ -0,0 +1,342 @@
...
+  protected pendingMessages: ChatMessage[] = [];
+  protected pendingRelevantContexts: Context[] = [];
+  protected isChatPageDisplayed = false;
+
+  constructor(
+    protected readonly context: ExtensionContext,
+    protected readonly agent: Agent,
+    protected readonly logger: LogOutputChannel,
+    protected readonly gitProvider: GitProvider,
+  ) {}
```
Comment (wsxiaoys) [inline clients/vscode/src/chat/BaseChatView.ts:21]:
```suggestion
  client?: ServerApi;
  protected pendingMessages: ChatMessage[] = [];
  protected pendingRelevantContexts: Context[] = [];
  protected isChatPageDisplayed = false;

  constructor(
    webview: Webview,
    protected readonly context: ExtensionContext,
    protected readonly agent: Agent,
    protected readonly logger: LogOutputChannel,
    protected readonly gitProvider: GitProvider,
  ) {}
```

Comment (wsxiaoys) [reply clients/vscode/src/chat/BaseChatView.ts:21]: consider change all `protected` to private, and expose method if needed.

File: clients/vscode/src/chat/BaseChatView.ts [hunk 3]
```diff
@@ -0,0 +1,342 @@
...
+import type { AgentFeature as Agent } from "../lsp/AgentFeature";
+import { GitProvider } from "../git/GitProvider";
+
+export abstract class BaseChatView {
+  webview?: WebviewView | WebviewPanel;
+  client?: ServerApi;
+  protected pendingMessages: ChatMessage[] = [];
+  protected pendingRelevantContexts: Context[] = [];
+  protected isChatPageDisplayed = false;
+
```
Comment (wsxiaoys) [inline clients/vscode/src/chat/BaseChatView.ts:15]: function initializeWebview(...) {
   this.webview = ...;
   this.webview.options = {...}
}

File: clients/vscode/src/chat/BaseChatView.ts
Comment (wsxiaoys) [inline clients/vscode/src/chat/BaseChatView.ts:1]: rename to `WebviewHelper`

File: clients/vscode/src/chat/chatPanel.ts (hunk 1 above)
Comment (antimonyGu) [reply clients/vscode/src/chat/chatPanel.ts:28]: I think we can do this refactor in a separate PR to make sure the refactor of `ChatSideViewProvider` as lean as possible. @wsxiaoys

**Pull request closed without merging**